domain/cloudimagemetadata/state/state.go

```
@@ -0,0 +1,354 @@
+func getArchitectureID(name string) (int, bool) {
+	id, ok := architectureIDsByName[name]
+	return id, ok
+}
+
+// SaveMetadata stores the provided list of cloud image metadata into the database.
+//
+// Returns [cloudmetadataerrors.Invalid] if the provided metadata omits required fields.
+//
+// It also fires a cleanup for old images, if any.
+//
+// [cloudimagemetadata.Metadata] are considered unique among a signature, composed of these fields:
+//
+//   - Stream
+//   - Region
+//   - Version
+//   - Arch
+//   - VirtType
+//   - RootStorageType
+//   - Source
+//
+// Above behaviors applies for duplicated inserted metadata:
+//   - If a metadata has the same signature than an existing one in the database, the imageID will be
+//     updated with the new value in the existing metadata
+//   - If a several metadatas have the same signature in the list, it will cause a unique constraint error from
+//     the database. It is likely a programmatic error.
+func (s *State) SaveMetadata(ctx context.Context, metadatas []cloudimagemetadata.Metadata) error {
+	// Cleaning the database before insert/update is a reasonable way to remove outdated data without
+	// requiring an housekeeping goroutine. Since a metadata data doesn't live long in the db and
+	// it shouldn't be a lot of custom metadata, it shouldn't slow too much the insertion.
+	// Expired metadata are ignored when retrieving metadata, so it doesn't have any impact to keep
+	// some of them longer than their time to live.
+	if err := s.tryCleanUpExpiredMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare inputs
+	createdAt := s.clock.Now()
+	values := make([]inputMetadata, 0, len(metadatas))
+	for _, m := range metadatas {
+		// Convert architecture name to a db id
+		archId, ok := getArchitectureID(m.Arch)
```

It will be expected that the `m.Arch` will be either validated or normalized in the service layer. This means, we can just return an error if it's not a valid architecture.

---

Ok, I will return an error and add to the service task to take care of it.

---

Thanks.